docs/v1/guidance-for-generating-p4info.md

```diff
@@ -0,0 +1,363 @@
+
+P4Info `MatchField`, `Param`, and probably soon also `Metadata`
+messages may optionally contain the following two fields:
+
++ `int32 bitwidth`
++ `P4NamedType type_name`
+
+Below we will describe what values these fields should have.
+
+Consider a single constrained value `x`.  Create a list of types
+`type_list(x)` using the pseudocode shown below.
+
+```
+type_list(x) {
+    tlist = [];    // tlist initialized to empty list
+    T = declared type of object x in the P4 program;
+    while (true) {
+        if (T is declared as "type B T") {
+            tlist = tlist + [T];   // append T to end of tlist
+            T = B;
+        } else if (T is declared as "typedef B T") {
+            T = B;
+        } else {
+            tlist = tlist + [T];   // append T to end of tlist
+            return tlist;
+        }
+    }
+}
+```
+
+Note that `type_list(x)` always starts with zero or more `type` names,
+and always ends with one type that is neither a `type` nor `typedef`
+name, e.g. `bit<W>`, a header type, struct type, etc.  It never
+contains the name of a type declared using `typedef`.  P4Runtime v1.0
+only supports `p4runtime_translation` annotations on `type`
+definitions.  If any such annotations occur on a `typedef` definition,
+they should be ignored.
+
+The `p4c` compiler signals an error if you attempt to create a cycle
+of type names.  In order to create such a cycle, the first `type` or
+`typedef` that appears in the program would have to refer to a later
+type name, and this is not allowed.
+
+If the last type is not `bit<W>` or `enum bit<W>`, that is an error
+for P4Runtime v1.0.  The "base" type must always be one of those for
+every constrained value.
+
+
+### `type_name` field
+
+Let `first_type` be the first element of the list `type_list(x)`.
+
+If `first_type` is a `type` name (i.e. not `bit<W>` or `enum bit<W>`),
+then the value of the P4Info `type_name` field should be `{name =
+"first_type_name"}`, where `first_type_name` is the name of
+`first_type`.
+
+Otherwise, the `type_name` field should be unset in the P4info
+message.
+
+
+### `bitwidth` field
+
+If `first_type` is a `type` name, _and_ if the `type` definition for
+this type has a `p4runtime_translation(uri_string, n)` annotation in
+the source code, then the P4Info `bitwidth` field should be assigned
+the value `n` that is the second parameter of that
+`p4runtime_translation` annotation.
+
+Otherwise, `bitwidth` should be equal to `W` where `bit<W>` or `enum
+bit<W>` is the last element of `type_list(x)`.
+
+Note that all type names that are present as the value of a
+`type_name` field anywhere in a P4Info message must contain an entry
+describing it in the `type_info` field of the P4Info message.  Thus
+this `bitwidth` value could be considered redundant information, since
+it can be derived from the description of the type contained within
+the `type_info` field (see [this
+comment](https://github.com/p4lang/p4runtime/issues/189#issuecomment-471240515)).
```

Comment:
I never replied to this comment. I agree that it is redundant information, but I believe it is a good thing to have here. I don't think we want some match fields / params to have a "valid" `bitwidth` field and others to have an "unset" `bitwidth` field. Keeping things uniform gives "cleaner" implementations, and probably makes the P4Info message easier to read IMO.

Reply:
I have replaced the text "(see [this comment](https...))" with:

"However, it is considered worth keeping this
small amount of redundancy for simplicity of interpreting the messages
that contain these `bitwidth` fields, without having to find the value
elsewhere in the P4Info message."